src/coins.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::errors::Error;
+use serde::{Deserialize, Serialize};
+use std::result;
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    str::FromStr,
+};
+
+/// A specialised `Result` type for safecoin.
+pub type Result<T> = result::Result<T, Error>;
+
+/// The conversion from coin to raw value
+pub const COIN_TO_RAW_POWER_OF_10_CONVERSION: u32 = 9;
+
+/// The conversion from coin to raw value
+pub const COIN_TO_RAW_CONVERSION: u64 = 1_000_000_000;
+
+/// The maximum amount of safecoin represented by a single `Coins`.
+pub const MAX_COINS_VALUE: Coins =
+    Coins((u32::max_value() as u64 + 1) * COIN_TO_RAW_CONVERSION - 1);
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+/// Structure representing a safecoin amount.
+pub struct Coins(u64);
+
+impl Coins {
+    /// New value from a number of nano coin.
+    pub fn from_nano(value: u64) -> Result<Self> {
+        if value > MAX_COINS_VALUE.0 {
+            return Err(Error::ExcessiveValue);
+        }
+        Ok(Self(value))
+    }
+
+    /// The maximum value a `Coins` can represent.
+    pub fn max_value() -> Self {
+        MAX_COINS_VALUE
+    }
+
+    /// Total coin expressed in number of nano coin.
+    pub fn as_nano(self) -> u64 {
+        self.0
+    }
+
+    // Computes self + rhs, returning None if overflow occurred.
+    pub fn checked_add(self, rhs: Coins) -> Option<Coins> {
+        self.0
+            .checked_add(rhs.0)
+            .and_then(|nano| Coins::from_nano(nano).ok())
```

Comment:
I guess returning an `Option` is consistent with the built in types' `checked_xxx` operations, but it's a bit inconsistent with `Coins::from_nano()` and `Coins::from_str()` which return `Err(Error::ExcessiveValue)` for this case.